games/tictactoe/src/game.rs

```diff
@@ -0,0 +1,170 @@
+        let serialized = String::from_utf8(state).expect("string from varible value");
+        let game: Game = serde_json::from_str(serialized.as_str())?;
+        Ok(game)
+    }
+
+    fn save(&mut self, game: &Game) -> Result<(), Error> {
+        let serialized = serde_json::to_string(game)?;
+        set_var!("game_state", "{}", serialized)?;
+        Ok(())
+    }
+}
+
+impl Game {
+    pub fn new(player_ids: Vec<String>) -> Self {
+        let mut board: Vec<String> = vec![];
+        board.resize(9, "".into());
+        return Game {
+            board: board,
+            current_player: player_ids[0].to_string(),
+            state:GameState::Playing,
+            version: 0,
+            player_ids,
+        };
+    }
+
+    pub fn render(&self, assigns: Assigns) -> String {
+        match self.state {
+            GameState::Playing => self.render_board(assigns),
+            GameState::Ended => self.render_board(assigns),
+        }
+    }
+
+    fn render_board(&self, assigns: Assigns) -> String {
+        let mut context = tera::Context::new();
+        context.insert("css", std::str::from_utf8(APP_CSS).unwrap());
+        context.insert("current_player", self.current_player.as_str());
+        context.insert("player_id", assigns.player_id.as_str());
+        context.insert("board", &self.board);
+        Tera::one_off(std::str::from_utf8(APP_HTML).unwrap(), &context, false).unwrap()
+    }
+
+    pub fn current_player_character(&self) -> String {
+        if self.player_ids[0] == self.current_player {
+            return "X".into();
+        }
+        return "O".into();
+    }
+
+    pub fn moved(&mut self) {
+        if self.player_ids[0] == self.current_player {
+            self.current_player = self.player_ids[1].clone();
+        } else {
+            self.current_player = self.player_ids[0].clone();
+        }
+    }
+
+    pub fn inc_version(&mut self) {
+        self.version += 1;
+    }
+
+    pub fn error(&mut self, msg: String) -> Assigns {
+        self.version += 1;
+        //return AssignsBuilder::default().error(Some(msg)).build().unwrap();
+        return Assigns {
+            player_id: "".into(),
+            error: Some(msg),
+            version: self.version,
+        };
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    pub struct FakeStorage {
+        pub state: String,
+    }
+
+    impl FakeStorage {
+        pub fn new() -> Self { FakeStorage{state: "".into()} }
+    }
+
+    impl Persister for FakeStorage {
+        fn load(&self) -> Result<Game, Error> {
+            let game: Game = serde_json::from_str(&self.state.as_str())?;
+            Ok(game)
+        }
+
+        fn save(&mut self, game: &Game) -> Result<(), Error> {
+            let serialized = serde_json::to_string(game)?;
+            self.state = serialized;
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn run_simulation() -> Result<(), Error> {
```

Comment:
will have some real tests soon. this is just enough to help me not have to test in the elixir app anymore.